src/diffmaps.jl

```diff
@@ -46,33 +46,60 @@ end
 Fit a isometric mapping model to `data`.
 
 # Arguments
-* `data`: a matrix of observations. Each column of `data` is an observation.
+* `data::Matrix`: a (n_features, n_observations) matrix of observations. Each column of `data` is an observation.
+  if `isnothing(kernel)`, `data` is instead the (n_observations, n_observations) precomputed Gram matrix.
 
 # Keyword arguments
-* `maxoutdim`: a dimension of the reduced space.
-* `t`: a number of transitions
-* `α`: a normalization parameter
-* `ɛ`: a Gaussian kernel variance (the scale parameter)
+* `kernel::Union{Nothing, Function}=(x, y) -> exp(-sum((x .- y) .^ 2) / ɛ)`: the kernel function. 
+ maps two input vectors (observations) to a scalar (a metric of their similarity).
+ by default, a Gaussian kernel. if `isnothing(kernel)`, we assume `data` is instead 
+ the (n_observations, n_observations) precomputed Gram matrix.
+* `ɛ::Real=1.0`: the Gaussian kernel variance (the scale parameter). ignored if custom `kernel` passed.
+* `maxoutdim::Int=2`: the dimension of the reduced space.
+* `t::Int=1`: the number of transitions
+* `α::Real=0.0`: a normalization parameter
 
 # Examples
 ```julia
-M = fit(DiffMap, rand(3,100)) # construct diffusion map model
-R = transform(M)              # perform dimensionality reduction
+X = rand(3, 100)     # toy data matrix, 100 observations
+
+# default kernel
+M = fit(DiffMap, X)  # construct diffusion map model
+R = transform(M)     # perform dimensionality reduction
+
+# custom kernel
+kernel = (x, y) -> x' * y # linear kernel
+M = fit(DiffMap, X, kernel=kernel)
+
+# precomputed Gram matrix
+kernel = (x, y) -> x' * y # linear kernel
+K = StatsBase.pairwise(kernel, eachcol(X), symmetric=true) # custom Gram matrix
+M = fit(DiffMap, K, kernel=nothing)
 ```
 """
-function fit(::Type{DiffMap}, X::AbstractMatrix{T}; maxoutdim::Int=2, t::Int=1, α::Real=0.0, ɛ::Real=1.0) where {T<:Real}
-    # compute kernel matrix
-    sumX = sum(X.^ 2, dims=1)
-    L = exp.(-( transpose(sumX) .+ sumX .- 2*transpose(X) * X ) ./ convert(T, ɛ))
-    # L = pairwise((x,y)-> exp(-norm(x-y,2)^2/ε), Xtr)
+function fit(::Type{DiffMap}, X::AbstractMatrix{T};
+             ɛ::Real=1.0,
+             kernel::Union{Nothing, Function}=(x, y) -> exp(-sum((x .- y) .^ 2) / ɛ), 
+             maxoutdim::Int=2, 
+             t::Int=1, 
+             α::Real=0.0
+            ) where {T<:Real}
+    if isa(kernel, Function)
+        # compute Gram matrix
+        L = StatsBase.pairwise(kernel, eachcol(X), symmetric=true)
```

Comment:
You can import `pairwise`, and drop module name.

Reply:
actually, I think this is needed. this is not implemented in the version of `MultivariateStats` pinned by `ManifoldLearning`. how to proceed? update the `Project.toml` or keep as `StatsBase.`?

```julia
julia> MultivariateStats.pairwise(kernel, eachcol(X), symmetric=true)
ERROR: MethodError: no method matching pairwise(::var"#1#2", ::Base.Generator{Base.OneTo{Int64}, Base.var"#227#228"{Matrix{Float64}}}; symmetric=true)
Closest candidates are:
  pairwise(::Function, ::AbstractVecOrMat{T}) where T<:AbstractFloat at ~/.julia/packages/MultivariateStats/HTpHt/src/common.jl:150 got unsupported keyword argument "symmetric"
  pairwise(::Function, ::AbstractVecOrMat{T}, ::AbstractVecOrMat{T}) where T<:AbstractFloat at ~/.julia/packages/MultivariateStats/HTpHt/src/common.jl:143 got unsupported keyword argument "symmetric"
Stacktrace:
 [1] top-level scope
   @ REPL[5]:1
```

Reply:
bumping up `MultivariateStats` to `v0.9` causes the nearest neighbors tests to fail :(
so should we keep as `StatsBase.pairwise` for now?

Reply:
You should import `pairwise` from StatsBase, it's not longer defined in `MultivariateStats`

Reply:
`import StatsAPI: pairwise` fails to run:
```julia
X = rand(30,40)
kernel=(x,y)->x'*y
MultivariateStats.pairwise(kernel, eachcol(X), symmetric=true)
```
but if I use `import StatsBase: pairwise` the diffusion map tests run (woo-hoo) but then the nearest neighbor tests fail. could it be this line? 
https://github.com/wildart/ManifoldLearning.jl/blob/master/src/nearestneighbors.jl#L32

Reply:
As I mentioned `pairwise` is no longer defined in `MultivariateStats`. You should import `pairwise` from `StatsBase` for now.

Reply:
ok, I think it's good to merge then since I'm using the StatsBase.pairwise for the diffusion map.
if I `import StatsBase: pairwise` the nearest neighbor tests no longer run. so seems it relies on a different signature call of `pairwise`.
let me know if I'm mis-understanding or if there is anything else I can do to make this PR suitable for merging.

Reply:
Could you try `pairwise((x,y)->norm(x-y), eachcol(NN.fitted), eachcol(X))` for nearest neighbor instead of
https://github.com/wildart/ManifoldLearning.jl/blob/417bbdb5b1074844423826f85513604903405c0e/src/nearestneighbors.jl#L32